File: Java/lab19_Observer/task_3_1/src/FileReader.java
import java.io.*;

public class FileReader {

    private final String filePath;

    public FileReader(String filePath) {
        this.filePath = filePath;

        Thread thread = new Thread(this::read);
        thread.start();
    }

    public void read() {
        try (BufferedReader reader = new BufferedReader(new java.io.FileReader(filePath))) {
            reader.lines().forEach(this::onLineRead);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    void onLineRead(String line) {
        System.out.println(line);

    }


}

File: Cpp/lab19_Observer/task_3_1/src/file_reader.cpp
#include <thread>
#include <iostream>
#include <fstream>

#include "file_reader.hpp"


void FileReader::onLineRead(std::string const& line)
{
  std::cout << line << std::endl;
}

void FileReader::read()
{
  std::string line;

  try
  {
    std::ifstream reader(this->filePath, std::ios_base::in);
    while (std::getline(reader, line))
    {
      this->onLineRead(line);
    }
  }
  catch(std::string msg)
  {
    std::cerr << msg << std::endl;
  }
}

FileReader::FileReader(std::string const& filePath)
: filePath(filePath)
{
  std::thread thd(&FileReader::read, this);
  thd.detach();
}
